### PR TITLE
Backport TFormula version 13 and TF1 version 12 changes from master to 6.22

### PR DESCRIPTION
**Pull request by mrodozov**
This is a request for backport of https://github.com/root-project/root/pull/7035
to 6-22 patches because of this ClassDef change:
`   ClassDef(TF1, 11) // The Parametric 1-D function` (previous was 10)
which we are checking in this test:
https://github.com/cms-sw/cmssw/blob/master/DQMServices/Components/test/testSchemaEvolution.cpp#L44
and should have one or the other version, and we better use this fix in 6-22 anyway.

### CODE REVIEW

**Comment by mrodozov** [comment]
dears, any comment on this please ?

**Comment by couet** [comment]
@lmoneta is the expert in this field. He will take care.

**Review by lmoneta — COMMENTED** [review]
The PR looks fine to me. But it is not only a back port of the fix in TF1::Copy ( #7183), 
it contains also the improvements in TF1::GetRandom ( PR #6257).
I think that should be  separate PR if needed, but for consistency it should include also the similar changes in TH1.

**Comment by pcanal** [comment]
@phsft-bot build with flags -DCTEST_TEST_EXCLUDE_NONE=On

**Comment by mrodozov** [comment]
I added the TH1 changes, this leads to a lot of warnings,
I guess this comes from implementations expecting the old interfaces.
Should I keep the old interfaces as well so whatever was implementing them to "see" what's expected ? (excuse my ignorance but I haven't been coding consistently for some time to remember this much by heart)

**Comment by mrodozov** [comment]
never mind the last question if the old ifaces are kept this leads to ambiguity and it doesn't compile

**Comment by mrodozov** [comment]
@lmoneta did I understand correct how to get also the TH1 changes ?

**Comment by lmoneta** [comment]
Hi, 
Do you need to have the changes in GetRandom in 6.22 ? 
These are more feature changes than bug fixes, so I will prefer avoiding the back port.
Anyway, yes this should be in case a separate PR to distinguish the two cases. 
Furthermore, if we back port the fix in TF1::Copy we should also back port this one too, 
https://github.com/root-project/root/pull/7274.

If you have trouble doing this back port, I can do it. But please confirm, what exactly you need to be back ported in 6.22

Cheers 
 Lorenzo

**Comment by mrodozov** [comment]
I'd prefer avoiding it, too. All I need is the changes in TFormula and it's ClassDef 
change:

https://github.com/root-project/root/pull/7193/files#diff-ca4416e47d798b9a8d5fbf80c478df1437d0ea6a9def18939f8d9c6d29869906R254

If I take line 96 and 254 alone, where I assume 96 is the reason for incrementing the ClassDef 
will that be enough ? I need the TFormula ClassDef change, thats it

**Comment by mrodozov** [comment]
@lmoneta I took only the TFormula changes that I need 
`std::unique_ptr<TMethodCall> fMethod`
this avoids the other backports

**Comment by lmoneta** [comment]
Yes, this back port should be fine, also fMethod is transient so there will be no issue in the I/O. 
If you need the back port for the TF1 (memory leak fixes in TF1 Copy) we can add as a separate PR
I will run the CI before merging it 
Thank you for updating this PR

**Comment by lmoneta** [comment]
@phsft-bot build with flags -DCTEST_TEST_EXCLUDE_NONE=On

**Comment by mrodozov** [comment]
No, we don't need TF1 changes. Only the TFormula ClassDef version to be consistent in 622 and master
We'll get the TF1 fix with 6.24

**Comment by mrodozov** [comment]
Sorry this is all wrong . We need the TF1 in ClassDef in fact. I got confused of all this adjustments :/

**Comment by mrodozov** [comment]
@lmoneta now it's fine . I got the changes also for TF1 without touching the GetRandom changes (just the pointer modifications as in TFormula)
 . sorry for the noise 😅 I'll squeeze the commits if necessary

**Comment by lmoneta** [comment]
Yes, I think is ok now.
Maybe it would be better and more consitent to backport directly the version 12 of TF1 existing in master, which includes  this recent commit:
https://github.com/root-project/root/pull/7274/commits/6a70887d3e61f71070a1ab203810e751b1db5eae

**Comment by mrodozov** [comment]
it's updated to the last version

**Comment by lmoneta** [comment]
Thank you ! Looks good to me now. 
Before merging, let;s see the CI results

**Comment by lmoneta** [comment]
@phsft-bot build with flags -DCTEST_TEST_EXCLUDE_NONE=On

**Comment by mrodozov** [comment]
I see the test values for stress test were increased in the same PR
https://github.com/root-project/root/pull/7274/commits/fe5c74b5ed6045eb72091bad693f7af1fdf2266e
but I don't see why the other test is failing :/

**Comment by mrodozov** [comment]
@lmoneta can I do something to fix this test ?

**Comment by lmoneta** [comment]
Hi, 
Yes this PR needs to include the changes in the stresss test of fe5c74b.
I will check and try to understand why the other test is failing

**Comment by mrodozov** [comment]
limits are updated

File: hist/hist/inc/TF1.h [hunk 1]
```diff
@@ -258,22 +257,23 @@ class TF1 : public TNamed, public TAttLine, public TAttFill, public TAttMarker {
...
+   std::unique_ptr<TMethodCall> fMethodCall; //!Pointer to MethodCall in case of interpreted function
    Bool_t      fNormalized{false};    //Normalization option (false by default)
    Double_t    fNormIntegral{};        //Integral of the function before being normalized
-   TF1FunctorPointer  *fFunctor{nullptr}; //! Functor object to wrap any C++ callable object
-   TFormula    *fFormula{nullptr};        //Pointer to TFormula in case when user define formula
-   TF1Parameters *fParams{nullptr};   //Pointer to Function parameters object (exists only for not-formula functions)
+   std::unique_ptr<TF1FunctorPointer>  fFunctor; //! Functor object to wrap any C++ callable object 
+   std::unique_ptr<TFormula>   fFormula;        //Pointer to TFormula in case when user define formula 
+   std::unique_ptr<TF1Parameters> fParams;   //Pointer to Function parameters object (exists only for not-formula functions)
    std::unique_ptr<TF1AbsComposition> fComposition; //! Pointer to composition (NSUM or CONV)
```
Comment (lmoneta) [inline hist/hist/inc/TF1.h:266]: This should not be a transient member!! 

Please remove the `!` and change from 
```
std::unique_ptr<TF1AbsComposition> fComposition; //! Pointer to composition (NSUM or CONV)
```
to 
```
std::unique_ptr<TF1AbsComposition> fComposition; // Pointer to composition (NSUM or CONV)
```

**Review by lmoneta — CHANGES_REQUESTED** [review]

File: hist/hist/inc/TF1.h (hunk 1 above)
Comment (lmoneta) [reply hist/hist/inc/TF1.h:266]: I think this caused the failare observed in the TF1 test. 
Thank you for updating it. I will run then the builds again afterwards

**Comment by mrodozov** [comment]
ok it's updated

**Comment by lmoneta** [comment]
Thank you!

**Comment by lmoneta** [comment]
@phsft-bot build with flags -DCTEST_TEST_EXCLUDE_NONE=On

**Pull request merged**